File: Reference/src/Brontide.Reference.Studio/ProviderPublisherTrustPolicyRegistry.cs
using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;

namespace Brontide.Reference.Studio;

public readonly record struct ProviderPublisherTrustPolicyAuthorityId
{
    private ProviderPublisherTrustPolicyAuthorityId(string value) => Value = value;
    public string Value { get; }

    public static ProviderPublisherTrustPolicyAuthorityId Create(string value)
    {
        ArgumentNullException.ThrowIfNull(value);
        if (!ContentAddressedProviderStore.IsDigest(value))
        {
            throw new ArgumentException("A policy authority identity must be an uppercase SHA-256 digest.", nameof(value));
        }
        return new(value);
    }
}

public sealed record ProviderPublisherTrustPolicyUpdate(
    long Sequence,
    ProviderPublisherTrustPolicyId? PreviousPolicyIdentity,
    ProviderPublisherTrustPolicy Policy,
    string Algorithm,
    string AuthorityPublicKeySpkiBase64,
    string SignatureBase64);

/// <summary>
/// One CBI57 authority transition. It carries both signatures because the predecessor authorizes the
/// transition and the successor proves its own key exists; a statement is meaningless without both.
/// </summary>
public sealed record ProviderPolicyAuthorityRotationStatement(
    long Generation,
    long PolicySequence,
    ProviderPublisherTrustPolicyId? PolicyIdentity,
    ProviderPublisherTrustPolicyAuthorityId PreviousAuthority,
    ProviderPublisherTrustPolicyAuthorityId NextAuthority,
    string Algorithm,
    string PreviousAuthorityPublicKeySpkiBase64,
    string NextAuthorityPublicKeySpkiBase64,
    string PreviousSignatureBase64,
    string NextSignatureBase64);

public static class ProviderPolicyAuthorityRotationManifest
{
    public static byte[] Encode(
        long generation,
        long policySequence,
        ProviderPublisherTrustPolicyId? policyIdentity,
        ProviderPublisherTrustPolicyAuthorityId previousAuthority,
        ProviderPublisherTrustPolicyAuthorityId nextAuthority)
    {
        if (generation <= 0 || policySequence < 0)
            throw new ArgumentException("A valid policy authority transition is required.");
        using var output = new MemoryStream();
        Append(output, "CBI57");
        Append(output, generation);
        Append(output, policySequence);
        Append(output, policyIdentity.HasValue ? 1 : 0);
        if (policyIdentity.HasValue) Append(output, policyIdentity.Value.Value);
        Append(output, previousAuthority.Value);
        Append(output, nextAuthority.Value);
        return output.ToArray();
    }

    private static void Append(Stream output, int value)
    {
        Span<byte> bytes = stackalloc byte[sizeof(int)];
        BinaryPrimitives.WriteInt32BigEndian(bytes, value);
        output.Write(bytes);
    }

    private static void Append(Stream output, long value)
    {
        Span<byte> bytes = stackalloc byte[sizeof(long)];
        BinaryPrimitives.WriteInt64BigEndian(bytes, value);
        output.Write(bytes);
    }

    private static void Append(Stream output, string value)
    {
        var bytes = Encoding.UTF8.GetBytes(value);
        Append(output, bytes.Length);
        output.Write(bytes);
    }
}

public sealed record ProviderPolicyAuthorityRotationResult(
    string Code,
    long Generation,
    ProviderPublisherTrustPolicyAuthorityId ActiveAuthority)
{
    public bool IsApplied => Code == "policy-authority-rotation-applied";
}

public static class ProviderPublisherTrustPolicyUpdateManifest
{
    public static byte[] Encode(
        long sequence,
        ProviderPublisherTrustPolicyId? previousPolicyIdentity,
        ProviderPublisherTrustPolicyId policyIdentity)
    {
        if (sequence <= 0) throw new ArgumentOutOfRangeException(nameof(sequence));
        using var output = new MemoryStream();
        Append(output, "CBI37");
        Append(output, sequence);
        Append(output, previousPolicyIdentity.HasValue ? 1 : 0);
        if (previousPolicyIdentity.HasValue) Append(output, previousPolicyIdentity.Value.Value);
        Append(output, policyIdentity.Value);
        return output.ToArray();
    }

    public static string Digest(
        long sequence,
        ProviderPublisherTrustPolicyId? previousPolicyIdentity,
        ProviderPublisherTrustPolicyId policyIdentity) =>
        Convert.ToHexString(SHA256.HashData(Encode(sequence, previousPolicyIdentity, policyIdentity)));

    private static void Append(Stream output, int value)
    {
        Span<byte> bytes = stackalloc byte[sizeof(int)];
        BinaryPrimitives.WriteInt32BigEndian(bytes, value);
        output.Write(bytes);
    }

    private static void Append(Stream output, long value)
    {
        Span<byte> bytes = stackalloc byte[sizeof(long)];
        BinaryPrimitives.WriteInt64BigEndian(bytes, value);
        output.Write(bytes);
    }

    private static void Append(Stream output, string value)
    {
        var bytes = Encoding.UTF8.GetBytes(value);
        Append(output, bytes.Length);
        output.Write(bytes);
    }
}

public sealed record VerifiedProviderPublisherTrustPolicySnapshot
{
    private VerifiedProviderPublisherTrustPolicySnapshot(
        ProviderPublisherTrustPolicyAuthorityId authorityIdentity,
        long sequence,
        ProviderPublisherTrustPolicy policy)
    {
        AuthorityIdentity = authorityIdentity;
        Sequence = sequence;
        Policy = policy;
    }

    public ProviderPublisherTrustPolicyAuthorityId AuthorityIdentity { get; }
    public long Sequence { get; }
    public ProviderPublisherTrustPolicy Policy { get; }

    internal static VerifiedProviderPublisherTrustPolicySnapshot Issue(
        ProviderPublisherTrustPolicyAuthorityId authorityIdentity,
        long sequence,
        ProviderPublisherTrustPolicy policy) => new(authorityIdentity, sequence, policy);
}

public sealed record ProviderPublisherTrustPolicyUpdateResult(
    string Code,
    VerifiedProviderPublisherTrustPolicySnapshot? Current)
{
    public bool IsApplied => Code == "policy-update-applied";
}

public sealed class ProviderPublisherTrustPolicyRegistry
{
    private readonly object _sync = new();
    private readonly ProviderPublisherTrustPolicyAuthorityId _authorityIdentity;
    private ProviderPublisherTrustPolicyAuthorityId _activeAuthority;
    private long _authorityGeneration;
    private VerifiedProviderPublisherTrustPolicySnapshot? _current;

    public ProviderPublisherTrustPolicyRegistry(ProviderPublisherTrustPolicyAuthorityId authorityIdentity)
    {
        _authorityIdentity = authorityIdentity;
        _activeAuthority = authorityIdentity;
    }

    /// <summary>
    /// The out-of-band pin. It never moves, which is what lets a chain recorded against it stay
    /// comparable across an authority rotation.
    /// </summary>
    internal ProviderPublisherTrustPolicyAuthorityId AuthorityIdentity => _authorityIdentity;

    /// <summary>The authority that may sign the next policy update, which a rotation moves.</summary>
    public ProviderPublisherTrustPolicyAuthorityId ActiveAuthorityIdentity
    {
        get { lock (_sync) return _activeAuthority; }
    }

    public long AuthorityGeneration
    {
        get { lock (_sync) return _authorityGeneration; }
    }

    public VerifiedProviderPublisherTrustPolicySnapshot? Current
    {
        get { lock (_sync) return _current; }
    }

    public ProviderPublisherTrustPolicyUpdateResult Apply(ProviderPublisherTrustPolicyUpdate update)
    {
        ArgumentNullException.ThrowIfNull(update);
        lock (_sync)
        {
            var code = Validate(update, out var policy);
            if (code is not null) return new(code, _current);

            if (_current is null)
            {
                if (update.Sequence != 1) return new("policy-update-sequence-invalid", null);
                if (update.PreviousPolicyIdentity.HasValue) return new("policy-update-predecessor-mismatch", null);
            }
            else
            {
                if (update.Sequence != _current.Sequence + 1) return new("policy-update-sequence-invalid", _current);
                if (update.PreviousPolicyIdentity != _current.Policy.Identity)
                    return new("policy-update-predecessor-mismatch", _current);
            }

            // The snapshot names the pin rather than the signing key: it is the trust root the policy
            // was verified under, and every downstream comparison of it must survive a rotation.
            _current = VerifiedProviderPublisherTrustPolicySnapshot.Issue(
                _authorityIdentity, update.Sequence, policy!);
            return new("policy-update-applied", _current);
        }
    }

    public ProviderPolicyAuthorityRotationResult Rotate(ProviderPolicyAuthorityRotationStatement statement)
    {
        ArgumentNullException.ThrowIfNull(statement);
        lock (_sync)
        {
            var code = ValidateRotation(statement);
            if (code is not null) return new(code, _authorityGeneration, _activeAuthority);
            _activeAuthority = statement.NextAuthority;
            _authorityGeneration = statement.Generation;
            return new("policy-authority-rotation-applied", _authorityGeneration, _activeAuthority);
        }
    }

    internal T WithCurrent<T>(Func<VerifiedProviderPublisherTrustPolicySnapshot?, T> action)
    {
        lock (_sync) return action(_current);
    }

    private string? ValidateRotation(ProviderPolicyAuthorityRotationStatement statement)
    {
        if (statement.Generation != _authorityGeneration + 1) return "policy-authority-generation-invalid";
        if (statement.PreviousAuthority != _activeAuthority) return "policy-authority-predecessor-mismatch";
        if (statement.NextAuthority == statement.PreviousAuthority) return "policy-authority-self-refused";
        if (statement.PolicySequence != (_current?.Sequence ?? 0)
            || statement.PolicyIdentity != _current?.Policy.Identity)
            return "policy-authority-chain-mismatch";
        if (statement.Algorithm != "ECDSA-P256-SHA256"
            || string.IsNullOrWhiteSpace(statement.PreviousAuthorityPublicKeySpkiBase64)
            || string.IsNullOrWhiteSpace(statement.NextAuthorityPublicKeySpkiBase64)
            || string.IsNullOrWhiteSpace(statement.PreviousSignatureBase64)
            || string.IsNullOrWhiteSpace(statement.NextSignatureBase64))
            return "policy-authority-evidence-invalid";
        try
        {
            var previousKey = Convert.FromBase64String(statement.PreviousAuthorityPublicKeySpkiBase64);
            var nextKey = Convert.FromBase64String(statement.NextAuthorityPublicKeySpkiBase64);
            if (Identify(previousKey) != statement.PreviousAuthority
                || Identify(nextKey) != statement.NextAuthority)
                return "policy-authority-key-mismatch";
            var manifest = ProviderPolicyAuthorityRotationManifest.Encode(
                statement.Generation, statement.PolicySequence, statement.PolicyIdentity,
                statement.PreviousAuthority, statement.NextAuthority);
            if (!TryVerify(previousKey, manifest, statement.PreviousSignatureBase64, out var previousValid)
                || !TryVerify(nextKey, manifest, statement.NextSignatureBase64, out var nextValid))
                return "policy-authority-evidence-invalid";
            if (!previousValid) return "policy-authority-signature-invalid";
            return nextValid ? null : "policy-authority-successor-unproven";
        }
        catch (Exception exception) when (exception is FormatException or CryptographicException or ArgumentException)
        {
            return "policy-authority-evidence-invalid";
        }
    }

    private static ProviderPublisherTrustPolicyAuthorityId Identify(byte[] publicKey) =>
        ProviderPublisherTrustPolicyAuthorityId.Create(Convert.ToHexString(SHA256.HashData(publicKey)));

    private static bool TryVerify(byte[] publicKey, byte[] manifest, string signatureBase64, out bool valid)
    {
        valid = false;
        var signature = Convert.FromBase64String(signatureBase64);
        using var verifier = ECDsa.Create();
        verifier.ImportSubjectPublicKeyInfo(publicKey, out var bytesRead);
        var parameters = verifier.ExportParameters(false);
        if (bytesRead != publicKey.Length || verifier.KeySize != 256
            || parameters.Curve.Oid.Value != "1.2.840.10045.3.1.7")
            return false;
        valid = verifier.VerifyData(manifest, signature, HashAlgorithmName.SHA256,
            DSASignatureFormat.Rfc3279DerSequence);
        return true;
    }

    private string? Validate(
        ProviderPublisherTrustPolicyUpdate update,
        out ProviderPublisherTrustPolicy? policy)
    {
        policy = null;
        if (update.Sequence <= 0 || update.Policy is null
            || !ProviderPublisherTrustEvaluator.TrySnapshot(update.Policy, out var snapshot))
            return "policy-update-policy-invalid";
        policy = snapshot;
        if (update.Algorithm != "ECDSA-P256-SHA256") return "policy-update-unsupported";
        try
        {
            var publicKey = Convert.FromBase64String(update.AuthorityPublicKeySpkiBase64);
            var signature = Convert.FromBase64String(update.SignatureBase64);
            if (Identify(publicKey) != _activeAuthority) return "policy-update-authority-mismatch";
            using var verifier = ECDsa.Create();
            verifier.ImportSubjectPublicKeyInfo(publicKey, out var bytesRead);
            var parameters = verifier.ExportParameters(false);
            if (bytesRead != publicKey.Length || verifier.KeySize != 256
                || parameters.Curve.Oid.Value != "1.2.840.10045.3.1.7")
                return "policy-update-malformed";
            return verifier.VerifyData(
                ProviderPublisherTrustPolicyUpdateManifest.Encode(
                    update.Sequence, update.PreviousPolicyIdentity, snapshot.Identity),
                signature, HashAlgorithmName.SHA256, DSASignatureFormat.Rfc3279DerSequence)
                ? null : "policy-update-signature-invalid";
        }
        catch (Exception exception) when (exception is FormatException or CryptographicException or ArgumentException)
        {
            return "policy-update-malformed";
        }
    }
}

public sealed class GovernedProviderArtifactAcquirer
{
    private readonly ProviderPublisherTrustPolicyRegistry _registry;
    private readonly TrustedProviderArtifactAcquirer _acquirer;

    public GovernedProviderArtifactAcquirer(
        ProviderPublisherTrustPolicyRegistry registry,
        TrustedProviderArtifactAcquirer acquirer)
    {
        ArgumentNullException.ThrowIfNull(registry);
        ArgumentNullException.ThrowIfNull(acquirer);
        _registry = registry;
        _acquirer = acquirer;
    }

    public TrustedProviderArtifactAcquisitionResult Acquire(
        ProviderArtifactAcquisitionRequest request,
        IProviderArtifactSource source,
        TrustedProviderPublisherAuthorization? authorization)
    {
        ArgumentNullException.ThrowIfNull(request);
        ArgumentNullException.ThrowIfNull(source);
        return _registry.WithCurrent(current =>
        {
            if (current is null)
                return TrustedProviderArtifactAcquisitionResult.Refused(
                    "publisher-trust-policy-unavailable", "publisher-evidence-not-evaluated", request.ExpectedSource);
            if (authorization is not null && authorization.PolicyIdentity != current.Policy.Identity)
                return TrustedProviderArtifactAcquisitionResult.Refused(
                    "publisher-authorization-superseded", "publisher-evidence-valid", request.ExpectedSource,
                    authorization.PolicyIdentity, authorization.PublisherKeyId);
            return _acquirer.Acquire(request, source, authorization);
        });
    }
}

File: Minimal/src/Brontide.Minimal.Host/ProviderPublisherTrustPolicyRegistry.fs
namespace Brontide.Minimal.Host

open System
open System.Buffers.Binary
open System.IO
open System.Security.Cryptography
open System.Text

[<StructuralEquality; StructuralComparison>]
type ProviderPublisherTrustPolicyAuthorityId = private ProviderPublisherTrustPolicyAuthorityId of string

[<RequireQualifiedAccess>]
module ProviderPublisherTrustPolicyAuthorityId =
    let create value =
        let valid =
            not (String.IsNullOrEmpty value) && value.Length = 64
            && value |> Seq.forall (fun c -> (c >= '0' && c <= '9') || (c >= 'A' && c <= 'F'))
        if not valid then invalidArg (nameof value) "A policy authority identity must be an uppercase SHA-256 digest."
        ProviderPublisherTrustPolicyAuthorityId value
    let value (ProviderPublisherTrustPolicyAuthorityId value) = value

type ProviderPublisherTrustPolicyUpdate =
    { Sequence: int64
      PreviousPolicyIdentity: ProviderPublisherTrustPolicyId option
      Policy: ProviderPublisherTrustPolicy
      Algorithm: string
      AuthorityPublicKeySpkiBase64: string
      SignatureBase64: string }

[<RequireQualifiedAccess>]
module ProviderPublisherTrustPolicyUpdateManifest =
    let private appendInt32 (output: Stream) value =
        let bytes = Array.zeroCreate<byte> sizeof<int>
        BinaryPrimitives.WriteInt32BigEndian(bytes.AsSpan(), value)
        output.Write bytes
    let private appendInt64 (output: Stream) value =
        let bytes = Array.zeroCreate<byte> sizeof<int64>
        BinaryPrimitives.WriteInt64BigEndian(bytes.AsSpan(), value)
        output.Write bytes
    let private appendString output (value: string) =
        let bytes = Encoding.UTF8.GetBytes value
        appendInt32 output bytes.Length
        output.Write bytes

    let encode sequence previous policyIdentity =
        if sequence <= 0L then invalidArg (nameof sequence) "A positive policy sequence is required."
        use output = new MemoryStream()
        appendString output "CBI37"
        appendInt64 output sequence
        appendInt32 output (if Option.isSome previous then 1 else 0)
        previous |> Option.iter (ProviderPublisherTrustPolicyId.value >> appendString output)
        appendString output (ProviderPublisherTrustPolicyId.value policyIdentity)
        output.ToArray()

    let digest sequence previous policyIdentity =
        encode sequence previous policyIdentity |> SHA256.HashData |> Convert.ToHexString

/// One CBI57 authority transition. It carries both signatures because the predecessor authorizes the
/// transition and the successor proves its own key exists; a statement is meaningless without both.
type ProviderPolicyAuthorityRotationStatement =
    { Generation: int64
      PolicySequence: int64
      PolicyIdentity: ProviderPublisherTrustPolicyId option
      PreviousAuthority: ProviderPublisherTrustPolicyAuthorityId
      NextAuthority: ProviderPublisherTrustPolicyAuthorityId
      Algorithm: string
      PreviousAuthorityPublicKeySpkiBase64: string
      NextAuthorityPublicKeySpkiBase64: string
      PreviousSignatureBase64: string
      NextSignatureBase64: string }

[<RequireQualifiedAccess>]
module ProviderPolicyAuthorityRotationManifest =
    let private appendInt32 (output: Stream) value =
        let bytes = Array.zeroCreate<byte> sizeof<int>
        BinaryPrimitives.WriteInt32BigEndian(bytes.AsSpan(), value)
        output.Write bytes
    let private appendInt64 (output: Stream) value =
        let bytes = Array.zeroCreate<byte> sizeof<int64>
        BinaryPrimitives.WriteInt64BigEndian(bytes.AsSpan(), value)
        output.Write bytes
    let private appendString output (value: string) =
        let bytes = Encoding.UTF8.GetBytes value
        appendInt32 output bytes.Length
        output.Write bytes

    let encode generation policySequence policyIdentity previousAuthority nextAuthority =
        if generation <= 0L || policySequence < 0L then
            invalidArg (nameof generation) "A valid policy authority transition is required."
        use output = new MemoryStream()
        appendString output "CBI57"
        appendInt64 output generation
        appendInt64 output policySequence
        appendInt32 output (if Option.isSome policyIdentity then 1 else 0)
        policyIdentity |> Option.iter (ProviderPublisherTrustPolicyId.value >> appendString output)
        appendString output (ProviderPublisherTrustPolicyAuthorityId.value previousAuthority)
        appendString output (ProviderPublisherTrustPolicyAuthorityId.value nextAuthority)
        output.ToArray()

type ProviderPolicyAuthorityRotationResult =
    { Code: string
      Generation: int64
      ActiveAuthority: ProviderPublisherTrustPolicyAuthorityId }
    member this.IsApplied = this.Code = "policy-authority-rotation-applied"

type VerifiedProviderPublisherTrustPolicySnapshot =
    private
    | VerifiedProviderPublisherTrustPolicySnapshot of
        ProviderPublisherTrustPolicyAuthorityId * int64 * ProviderPublisherTrustPolicy
    member this.AuthorityIdentity =
        let (VerifiedProviderPublisherTrustPolicySnapshot(authority, _, _)) = this
        authority
    member this.Sequence =
        let (VerifiedProviderPublisherTrustPolicySnapshot(_, sequence, _)) = this
        sequence
    member this.Policy =
        let (VerifiedProviderPublisherTrustPolicySnapshot(_, _, policy)) = this
        policy

type ProviderPublisherTrustPolicyUpdateResult =
    { Code: string
      Current: VerifiedProviderPublisherTrustPolicySnapshot option }
    member this.IsApplied = this.Code = "policy-update-applied"

type ProviderPublisherTrustPolicyRegistry(authorityIdentity: ProviderPublisherTrustPolicyAuthorityId) =
    let syncRoot = obj ()
    let mutable current: VerifiedProviderPublisherTrustPolicySnapshot option = None
    let mutable activeAuthority = authorityIdentity
    let mutable authorityGeneration = 0L

    let identify (publicKey: byte array) =
        SHA256.HashData publicKey |> Convert.ToHexString |> ProviderPublisherTrustPolicyAuthorityId.create

    /// Imports P-256 key material and reports whether the signature holds. `None` distinguishes key
    /// material that is not exactly P-256 SPKI from a signature that simply does not verify.
    let tryVerify (publicKey: byte array) (manifest: byte array) (signatureBase64: string) =
        let signature = Convert.FromBase64String signatureBase64
        use verifier = ECDsa.Create()
        let mutable bytesRead = 0
        verifier.ImportSubjectPublicKeyInfo(publicKey.AsSpan(), &bytesRead)
        let curveOid = verifier.ExportParameters(false).Curve.Oid.Value |> Option.ofObj
        if bytesRead <> publicKey.Length || verifier.KeySize <> 256 || curveOid <> Some "1.2.840.10045.3.1.7" then None
        else
            Some(verifier.VerifyData(
                manifest, signature, HashAlgorithmName.SHA256, DSASignatureFormat.Rfc3279DerSequence))

    let validateRotation (statement: ProviderPolicyAuthorityRotationStatement) =
        if statement.Generation <> authorityGeneration + 1L then Some "policy-authority-generation-invalid"
        elif statement.PreviousAuthority <> activeAuthority then Some "policy-authority-predecessor-mismatch"
        elif statement.NextAuthority = statement.PreviousAuthority then Some "policy-authority-self-refused"
        elif statement.PolicySequence <> (current |> Option.map _.Sequence |> Option.defaultValue 0L)
            || statement.PolicyIdentity <> (current |> Option.map _.Policy.Identity) then
            Some "policy-authority-chain-mismatch"
        elif statement.Algorithm <> "ECDSA-P256-SHA256"
            || String.IsNullOrWhiteSpace statement.PreviousAuthorityPublicKeySpkiBase64
            || String.IsNullOrWhiteSpace statement.NextAuthorityPublicKeySpkiBase64
            || String.IsNullOrWhiteSpace statement.PreviousSignatureBase64
            || String.IsNullOrWhiteSpace statement.NextSignatureBase64 then
            Some "policy-authority-evidence-invalid"
        else
            try
                let previousKey = Convert.FromBase64String statement.PreviousAuthorityPublicKeySpkiBase64
                let nextKey = Convert.FromBase64String statement.NextAuthorityPublicKeySpkiBase64
                if identify previousKey <> statement.PreviousAuthority
                    || identify nextKey <> statement.NextAuthority then Some "policy-authority-key-mismatch"
                else
                    let manifest =
                        ProviderPolicyAuthorityRotationManifest.encode statement.Generation
                            statement.PolicySequence statement.PolicyIdentity
                            statement.PreviousAuthority statement.NextAuthority
                    match tryVerify previousKey manifest statement.PreviousSignatureBase64,
                          tryVerify nextKey manifest statement.NextSignatureBase64 with
                    | None, _ | _, None -> Some "policy-authority-evidence-invalid"
                    | Some false, _ -> Some "policy-authority-signature-invalid"
                    | Some true, Some false -> Some "policy-authority-successor-unproven"
                    | Some true, Some true -> None
            with
            | :? FormatException
            | :? CryptographicException
            | :? ArgumentException -> Some "policy-authority-evidence-invalid"

    let validate (update: ProviderPublisherTrustPolicyUpdate) =
        match ProviderPublisherTrustEvaluator.tryValidate update.Policy with
        | None -> Error "policy-update-policy-invalid"
        | Some policy when update.Sequence <= 0L -> Error "policy-update-policy-invalid"
        | Some _ when update.Algorithm <> "ECDSA-P256-SHA256" -> Error "policy-update-unsupported"
        | Some policy ->
            try
                let publicKey = Convert.FromBase64String update.AuthorityPublicKeySpkiBase64
                let signature = Convert.FromBase64String update.SignatureBase64
                if identify publicKey <> activeAuthority then Error "policy-update-authority-mismatch"
                else
                    use verifier = ECDsa.Create()
                    let mutable bytesRead = 0
                    verifier.ImportSubjectPublicKeyInfo(publicKey.AsSpan(), &bytesRead)
                    let curveOid = verifier.ExportParameters(false).Curve.Oid.Value |> Option.ofObj
                    if bytesRead <> publicKey.Length || verifier.KeySize <> 256 || curveOid <> Some "1.2.840.10045.3.1.7" then
                        Error "policy-update-malformed"
                    elif verifier.VerifyData(
                        ProviderPublisherTrustPolicyUpdateManifest.encode
                            update.Sequence update.PreviousPolicyIdentity policy.Identity,
                        signature,
                        HashAlgorithmName.SHA256,
                        DSASignatureFormat.Rfc3279DerSequence) then Ok policy
                    else Error "policy-update-signature-invalid"
            with
            | :? FormatException
            | :? CryptographicException
            | :? ArgumentException -> Error "policy-update-malformed"

    member _.Current = lock syncRoot (fun () -> current)

    /// The out-of-band pin. It never moves, which is what lets a chain recorded against it stay
    /// comparable across an authority rotation.
    member internal _.AuthorityIdentity = authorityIdentity

    /// The authority that may sign the next policy update, which a rotation moves.
    member _.ActiveAuthorityIdentity = lock syncRoot (fun () -> activeAuthority)
    member _.AuthorityGeneration = lock syncRoot (fun () -> authorityGeneration)

    member _.Rotate(statement: ProviderPolicyAuthorityRotationStatement) =
        if isNull (box statement) then nullArg (nameof statement)
        lock syncRoot (fun () ->
            match validateRotation statement with
            | Some code -> { Code = code; Generation = authorityGeneration; ActiveAuthority = activeAuthority }
            | None ->
                activeAuthority <- statement.NextAuthority
                authorityGeneration <- statement.Generation
                { Code = "policy-authority-rotation-applied"
                  Generation = authorityGeneration
                  ActiveAuthority = activeAuthority })

    member _.Apply(update: ProviderPublisherTrustPolicyUpdate) =
        if isNull (box update) then nullArg (nameof update)
        lock syncRoot (fun () ->
            match validate update with
            | Error code -> { Code = code; Current = current }
            | Ok policy ->
                let chainCode =
                    match current with
                    | None when update.Sequence <> 1L -> Some "policy-update-sequence-invalid"
                    | None when Option.isSome update.PreviousPolicyIdentity -> Some "policy-update-predecessor-mismatch"
                    | Some snapshot when update.Sequence <> snapshot.Sequence + 1L -> Some "policy-update-sequence-invalid"
                    | Some snapshot when update.PreviousPolicyIdentity <> Some snapshot.Policy.Identity ->
                        Some "policy-update-predecessor-mismatch"
                    | _ -> None
                match chainCode with
                | Some code -> { Code = code; Current = current }
                | None ->
                    // The snapshot names the pin rather than the signing key: it is the trust root the
                    // policy was verified under, and every downstream comparison of it must survive a
                    // rotation.
                    let snapshot = VerifiedProviderPublisherTrustPolicySnapshot(authorityIdentity, update.Sequence, policy)
                    current <- Some snapshot
                    { Code = "policy-update-applied"; Current = current })

    member internal _.WithCurrent action = lock syncRoot (fun () -> action current)

type GovernedProviderArtifactAcquirer(
    registry: ProviderPublisherTrustPolicyRegistry,
    acquirer: TrustedProviderArtifactAcquirer) =
    do
        if isNull (box registry) then nullArg (nameof registry)
        if isNull (box acquirer) then nullArg (nameof acquirer)

    member _.Acquire(
        request: ProviderArtifactAcquisitionRequest,
        source: IProviderArtifactSource,
        authorization: TrustedProviderPublisherAuthorization option) =
        if isNull (box request) then nullArg (nameof request)
        if isNull (box source) then nullArg (nameof source)
        registry.WithCurrent(fun current ->
            match current with
            | None ->
                { TrustCode = "publisher-trust-policy-unavailable"
                  EvidenceCode = "publisher-evidence-not-evaluated"
                  PolicyIdentity = None
                  PublisherKeyId = None
                  SourceIdentity = request.ExpectedSource
                  TransportCode = "transport-not-attempted"
                  AdmissionCode = "admission-not-attempted"
                  Staged = None }
            | Some snapshot when authorization |> Option.exists (fun value -> value.PolicyIdentity <> snapshot.Policy.Identity) ->
                let supplied = authorization.Value
                { TrustCode = "publisher-authorization-superseded"
                  EvidenceCode = "publisher-evidence-valid"
                  PolicyIdentity = Some supplied.PolicyIdentity
                  PublisherKeyId = Some supplied.PublisherKeyId
                  SourceIdentity = request.ExpectedSource
                  TransportCode = "transport-not-attempted"
                  AdmissionCode = "admission-not-attempted"
                  Staged = None }
            | Some _ -> acquirer.Acquire(request, source, authorization))
